docs/architecture/adr-027-deterministic-protobuf-serialization.md

```diff
@@ -11,15 +12,34 @@ Proposed
 ## Context
 
 [Protobuf](https://developers.google.com/protocol-buffers/docs/proto3)
-seralization is not unique (i.e. there exist a practically unlimited number of
+serialization is not bijective (i.e. there exist a practically unlimited number of
 valid binary representations for a protobuf document)<sup>1</sup>. For signature
 verification in Cosmos SDK, signer and verifier need to agree on the same
 serialization of a SignDoc as defined in
 [ADR-020](./adr-020-protobuf-transaction-encoding.md) without transmitting the
-serialization. This document describes a deterministic serialization scheme for
+serialization. This document describes a bijective serialization scheme for
 a subset of protobuf documents, that covers this use case but can be reused in
 other cases as well.
 
+### Background - Protobuf3 Encoding
+
+All numeric types in protobuf3 are encoded as
+[varints](https://developers.google.com/protocol-buffers/docs/encoding#varints).
+Varints are at most 10 bytes, and since each varint byte has 7 bits of data,
+varints are a representation of `uint70` (70-bit unsigned integer). When
+encoding, numeric values are casted from their base type to `uint70`, and when
+decoding, the parsed `uint70` is casted to the appropriate numeric type.
```

Comment:
I think here's the root of all the confusion I had in my comments.

The casting you're describing seems like the _implementation_ of one encoder/decoder. Not a bad one, but also not the only one.

Here's another implementation, that I created specifically for this ADR, for decoding:
- read the field type given by the protobuf file
- if it's a uint32:
    - then in this case, I call my function `decode_varint_32`, which reads only up to 5 bytes
- works, because per this ADR, unit32 takes max 5 bytes.

This implementation never deals with 70-bit integers, and never casts.